docs/sql-ref-literals.md

```diff
@@ -0,0 +1,385 @@
+  +----------+
+{% endhighlight %}
+
+### Null Literal
+
+A null literal is used to specify a null value.
+
+#### <em>Syntax</em>
+
+{% highlight sql %}
+NULL
+{% endhighlight %}
+
+#### <em>Examples</em>
+
+{% highlight sql %}
+SELECT NULL AS col;
+  +----+
+  | col|
+  +----+
+  |NULL|
+  +----+
+{% endhighlight %}
+
+### Boolean Literal
+
+A boolean literal is used to specify a boolean value.
+
+#### <em>Syntax</em>
+
+{% highlight sql %}
+TRUE | FALSE
+{% endhighlight %}
+
+#### <em>Examples</em>
+
+{% highlight sql %}
+SELECT TRUE AS col;
+  +----+
+  | col|
+  +----+
+  |true|
+  +----+
+{% endhighlight %}
+
+### Numeric Literal
+
+A numeric literal is used to specify a fixed or floating-point number.
+
+#### Integer Literal
+
+#### <em>Syntax</em>
+
+{% highlight sql %}
+[ + | - ] digit [ ... ] [ L | S | Y ]
+{% endhighlight %}
+digit: one of 0, 1, 2, 3, 4, 5, 6, 7, 8 or 9. <br>
+default (no postfix): indicates a 4-byte signed integer number.<br>
+L: case insensitive, indicates <code>BIGINT</code>, which is a 8-byte signed integer number.<br>
+S: case insensitive, indicates <code>SMALLINT</code>, which is a 2-byte signed integer number.<br>
+Y: case insensitive, indicates <code>TINYINT</code>, which is a 1-byte signed integer number.<br>
+#### <em>Examples</em>
+
+{% highlight sql %}
+SELECT -2147483648 AS col;
+  +-----------+
+  |        col|
+  +-----------+
+  |-2147483648|
+  +-----------+
+
+SELECT 9223372036854775807l AS col;
+  +-------------------+
+  |                col|
+  +-------------------+
+  |9223372036854775807|
+  +-------------------+
+
+SELECT -32Y AS col;
+  +---+
+  |col|
+  +---+
+  |-32|
+  +---+
+
+SELECT 482S AS col;
+  +---+
+  |col|
+  +---+
+  |482|
+  +---+
+{% endhighlight %}
+
+#### Decimal Literal
+
+#### <em>Syntax</em>
+
+{% highlight sql %}
+ [ + | - ] { digit [ ... ] . [ digit [ ... ] ] | . digit [ ... ] }
+{% endhighlight %}
+
+#### <em>Examples</em>
+
+{% highlight sql %}
+SELECT 12.578 AS col;
+  +------+
+  |   col|
+  +------+
+  |12.578|
+  +------+
+
+SELECT -0.1234567 AS col;
+  +----------+
+  |       col|
+  +----------+
+  |-0.1234567|
+  +----------+
+
+SELECT -.1234567 AS col;
+  +----------+
+  |       col|
+  +----------+
+  |-0.1234567|
+  +----------+
+{% endhighlight %}
+
+#### Floating Point and BigDecimal Literals
+
+#### <em>Syntax</em>
+
+{% highlight sql %}
+ [ + | - ] { digit [ ... ] [ E [ + | - ] digit [ ... ] ] [ D | BD ]
+             | digit [ ... ] . [ digit [ ... ] ] [ E [ + | - ] digit [ ... ] ] [ D | BD ]
+             | . digit [ ... ] [ E [ + | - ] digit [ ... ] ] [ D | BD ] }
+{% endhighlight %}
+digit: one of 0, 1, 2, 3, 4, 5, 6, 7, 8 or 9. <br>
+default (no postfix): indicate a 4-byte single-precision floating point number.<br>
+D: case insensitive, indicates <code>DOUBLE</code>, which is a 8-byte double-precision floating point number.<br>
+BD: case insensitive, indicates <code>BIGDECIMAL</code>, which is an arbitrary-precision signed decimal number.<br>
```

Comment:
Where should we document some special literals? e.g., infinity, nan, ... https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L241-L247
```
scala> sql("select double('infinity')").show()
+------------------------+                                                      
|CAST(infinity AS DOUBLE)|
+------------------------+
|                Infinity|
+------------------------+
```

Reply:
I think strictly speaking, infinity, nan are not constants, because ```SELECT infinity``` doesn't work, so I guess no need to document these here.

Reply:
Where should we document them? I think we should do somewhere.

Reply:
I re-arranged data types page to add these floating point special values. Here is the PR https://github.com/apache/spark/pull/28264